File: php/OffsetOption.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

/**
 * OffsetOption - Specifies where to start consuming from.
 * Referencing Java OffsetOption.java
 */
class OffsetOption
{
    const TYPE_POLICY = 'POLICY';
    const TYPE_OFFSET = 'OFFSET';
    const TYPE_TAIL_N = 'TAIL_N';
    const TYPE_TIMESTAMP = 'TIMESTAMP';

    const POLICY_LAST_VALUE = 0;
    const POLICY_MIN_VALUE = 1;
    const POLICY_MAX_VALUE = 2;

    private $type;
    private $value;

    private function __construct(string $type, int $value)
    {
        $this->type = $type;
        $this->value = $value;
    }

    /**
     * Resume from last consumed position.
     */
    public static function lastOffset(): OffsetOption
    {
        return new self(self::TYPE_POLICY, self::POLICY_LAST_VALUE);
    }

    /**
     * Start from earliest offset.
     */
    public static function minOffset(): OffsetOption
    {
        return new self(self::TYPE_POLICY, self::POLICY_MIN_VALUE);
    }

    /**
     * Start from latest offset.
     */
    public static function maxOffset(): OffsetOption
    {
        return new self(self::TYPE_POLICY, self::POLICY_MAX_VALUE);
    }

    /**
     * Start from a specific offset.
     *
     * @param int $offset Must be >= 0
     */
    public static function ofOffset(int $offset): OffsetOption
    {
        if ($offset < 0) {
            throw new \InvalidArgumentException("Offset must be >= 0");
        }
        return new self(self::TYPE_OFFSET, $offset);
    }

    /**
     * Start from N messages from tail.
     *
     * @param int $n Must be >= 0
     */
    public static function ofTailN(int $n): OffsetOption
    {
        if ($n < 0) {
            throw new \InvalidArgumentException("Tail N must be >= 0");
        }
        return new self(self::TYPE_TAIL_N, $n);
    }

    /**
     * Start from messages at/after a timestamp.
     *
     * @param int $timestamp Unix timestamp in seconds, must be >= 0
     */
    public static function ofTimestamp(int $timestamp): OffsetOption
    {
        if ($timestamp < 0) {
            throw new \InvalidArgumentException("Timestamp must be >= 0");
        }
        return new self(self::TYPE_TIMESTAMP, $timestamp);
    }

    public function getType(): string
    {
        return $this->type;
    }

    public function getValue(): int
    {
        return $this->value;
    }

    public function isPolicy(): bool
    {
        return $this->type === self::TYPE_POLICY;
    }

    public function isOffset(): bool
    {
        return $this->type === self::TYPE_OFFSET;
    }

    public function isTailN(): bool
    {
        return $this->type === self::TYPE_TAIL_N;
    }

    public function isTimestamp(): bool
    {
        return $this->type === self::TYPE_TIMESTAMP;
    }
}

File: golang/offset_option.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golang

import (
	"fmt"

	v2 "github.com/apache/rocketmq-clients/golang/v5/protocol/v2"
)

type offsetOptionType int

const (
	offsetOptionTypePolicy offsetOptionType = iota
	offsetOptionTypeOffset
	offsetOptionTypeTailN
	offsetOptionTypeTimestamp
)

// OffsetOption specifies the starting offset for lite topic consumption.
type OffsetOption struct {
	optionType offsetOptionType
	value      int64
}

var (
	// LastOffset starts consuming from the last consumed offset of the consumer group.
	LastOffset = OffsetOption{optionType: offsetOptionTypePolicy, value: int64(v2.OffsetOption_LAST)}
	// MinOffset starts consuming from the minimum available offset.
	MinOffset = OffsetOption{optionType: offsetOptionTypePolicy, value: int64(v2.OffsetOption_MIN)}
	// MaxOffset starts consuming from the maximum available offset, skipping existing messages.
	MaxOffset = OffsetOption{optionType: offsetOptionTypePolicy, value: int64(v2.OffsetOption_MAX)}
)

// NewOffsetOptionWithOffset creates an OffsetOption from a specific offset.
func NewOffsetOptionWithOffset(offset int64) (OffsetOption, error) {
	if offset < 0 {
		return OffsetOption{}, fmt.Errorf("offset must be greater than or equal to 0")
	}
	return OffsetOption{optionType: offsetOptionTypeOffset, value: offset}, nil
}

// NewOffsetOptionWithTailN creates an OffsetOption from the last N messages.
func NewOffsetOptionWithTailN(tailN int64) (OffsetOption, error) {
	if tailN < 0 {
		return OffsetOption{}, fmt.Errorf("tailN must be greater than or equal to 0")
	}
	return OffsetOption{optionType: offsetOptionTypeTailN, value: tailN}, nil
}

// NewOffsetOptionWithTimestamp creates an OffsetOption from a Unix millisecond timestamp.
func NewOffsetOptionWithTimestamp(timestamp int64) (OffsetOption, error) {
	if timestamp < 0 {
		return OffsetOption{}, fmt.Errorf("timestamp must be greater than or equal to 0")
	}
	return OffsetOption{optionType: offsetOptionTypeTimestamp, value: timestamp}, nil
}

func (option OffsetOption) toProtobuf() *v2.OffsetOption {
	switch option.optionType {
	case offsetOptionTypePolicy:
		return &v2.OffsetOption{
			OffsetType: &v2.OffsetOption_Policy_{
				Policy: v2.OffsetOption_Policy(option.value),
			},
		}
	case offsetOptionTypeOffset:
		return &v2.OffsetOption{
			OffsetType: &v2.OffsetOption_Offset{
				Offset: option.value,
			},
		}
	case offsetOptionTypeTailN:
		return &v2.OffsetOption{
			OffsetType: &v2.OffsetOption_TailN{
				TailN: option.value,
			},
		}
	case offsetOptionTypeTimestamp:
		return &v2.OffsetOption{
			OffsetType: &v2.OffsetOption_Timestamp{
				Timestamp: option.value,
			},
		}
	default:
		return nil
	}
}
